Improved source validation

schedutron opened the pull request:
Since validator patterns are used repeatedly in a nested loop, it's better to compile them before for improved performance. Apart from these, at some places, source files were left open and not closed. I have fixed that security bug.

Lastly, I've tried to improve validator output, which till now gives redundant  and not much useful output when a format is violated.

Consider the old output:
```
12-23-2017 00:16:26 [<module>:222] INFO : Running source validator ...
12-23-2017 00:16:26 [<module>:223] INFO : Peloton root : /Users/binaryboy/Desktop/peloton
12-23-2017 00:16:26 [<module>:242] INFO : Scanning file : /Users/binaryboy/Desktop/peloton/src/main/peloton/peloton.cpp
12-23-2017 00:16:26 [check_common_patterns:112] INFO : Invalid pattern --  free\( -- found in : /Users/binaryboy/Desktop/peloton/src/main/peloton/peloton.cpp
12-23-2017 00:16:26 [check_common_patterns:113] INFO : Line #32 :: // free()

12-23-2017 00:16:26 [check_format:138] INFO : Invalid formatting in file : /Users/binaryboy/Desktop/peloton/src/main/peloton/peloton.cpp
12-23-2017 00:16:26 [check_format:138] INFO : Invalid formatting in file : /Users/binaryboy/Desktop/peloton/src/main/peloton/peloton.cpp
12-23-2017 00:16:26 [check_format:138] INFO : Invalid formatting in file : /Users/binaryboy/Desktop/peloton/src/main/peloton/peloton.cpp
12-23-2017 00:16:26 [check_format:138] INFO : Invalid formatting in file : /Users/binaryboy/Desktop/peloton/src/main/peloton/peloton.cpp
12-23-2017 00:16:26 [check_format:138] INFO : Invalid formatting in file : /Users/binaryboy/Desktop/peloton/src/main/peloton/peloton.cpp
12-23-2017 00:16:26 [check_format:138] INFO : Invalid formatting in file : /Users/binaryboy/Desktop/peloton/src/main/peloton/peloton.cpp
12-23-2017 00:16:26 [check_format:138] INFO : Invalid formatting in file : /Users/binaryboy/Desktop/peloton/src/main/peloton/peloton.cpp
12-23-2017 00:16:26 [check_format:138] INFO : Invalid formatting in file : /Users/binaryboy/Desktop/peloton/src/main/peloton/peloton.cpp
12-23-2017 00:16:26 [check_format:138] INFO : Invalid formatting in file : /Users/binaryboy/Desktop/peloton/src/main/peloton/peloton.cpp
12-23-2017 00:16:26 [check_format:138] INFO : Invalid formatting in file : /Users/binaryboy/Desktop/peloton/src/main/peloton/peloton.cpp
12-23-2017 00:16:26 [check_format:138] INFO : Invalid formatting in file : /Users/binaryboy/Desktop/peloton/src/main/peloton/peloton.cpp
12-23-2017 00:16:26 [check_format:138] INFO : Invalid formatting in file : /Users/binaryboy/Desktop/peloton/src/main/peloton/peloton.cpp
12-23-2017 00:16:26 [check_format:138] INFO : Invalid formatting in file : /Users/binaryboy/Desktop/peloton/src/main/peloton/peloton.cpp
12-23-2017 00:16:26 [check_format:138] INFO : Invalid formatting in file : /Users/binaryboy/Desktop/peloton/src/main/peloton/peloton.cpp
12-23-2017 00:16:26 [check_format:138] INFO : Invalid formatting in file : /Users/binaryboy/Desktop/peloton/src/main/peloton/peloton.cpp
12-23-2017 00:16:26 [<module>:245] INFO : Validation NOT successful
```

vs. the new one:
```
12-23-2017 00:38:47 [<module>:232] INFO : Running source validator ...
12-23-2017 00:38:47 [<module>:233] INFO : Peloton root : /Users/binaryboy/Desktop/peloton
12-23-2017 00:38:47 [<module>:252] INFO : Scanning file : /Users/binaryboy/Desktop/peloton/src/main/peloton/peloton.cpp
12-23-2017 00:38:47 [check_common_patterns:112] INFO : Invalid pattern --  free\( -- found in : /Users/binaryboy/Desktop/peloton/src/main/peloton/peloton.cpp
12-23-2017 00:38:47 [check_common_patterns:113] INFO : Line #32 :: // free()

12-23-2017 00:38:48 [check_format:148] INFO : Invalid formatting in file : /Users/binaryboy/Desktop/peloton/src/main/peloton/peloton.cpp
12-23-2017 00:38:48 [check_format:149] INFO : Line:     ::google::SetUsageMessage("Usage Info: \n");::google::HandleCommandLineHelpFlags();}

12-23-2017 00:38:48 [check_format:148] INFO : Invalid formatting in file : /Users/binaryboy/Desktop/peloton/src/main/peloton/peloton.cpp
12-23-2017 00:38:48 [check_format:149] INFO : Line: // free()

12-23-2017 00:38:48 [<module>:255] INFO : Validation NOT successful
``` 
for the same format anomaly.

apavlo commented:
@schedutron Thanks for sending this. @tcm-marcel will review this when he gets back from his trip in the woods.

tcm-marcel commented on script/validators/source_validator.py:
```diff
@@ -127,19 +129,29 @@ def check_format(file_path):
         diff = difflib.unified_diff(src, formatted_src)
-        for line in diff:
-            LOG.info("Invalid formatting in file : " + file_path)
+        if diff.next():  # Got a non-empty diff.
```
The interator is incremented when you call `.next()`, so I think you miss the first element in the later loop. Or does the loop reset the iterator?

tcm-marcel commented on script/validators/source_validator.py:
```diff
@@ -127,19 +129,29 @@ def check_format(file_path):
+                continue
+            if line.startswith('-'):
+                LOG.info("Invalid formatting in file : " + file_path)
+                LOG.info("Line: %s" % line[1:])
```
In my opinion the content of the line is not useful without line numbers. As we have a formatter script for that, you can leave it the old way. Or you find a way to display the line numbers here, I don't know if libdiff can do that.

tcm-marcel commented on .gitignore:
```diff
@@ -312,3 +312,25 @@ third_party/gflags/
 
 # Test output file logs to ignore
 stats_log
+
```
Please do not commit any changes to `.gitignore`.

tcm-marcel commented on script/validators/source_validator.py:
```diff
@@ -38,9 +38,11 @@
+    )
 
-CLANG_FORMAT = "clang-format"
+CLANG_FORMAT = "clang-format-3.7"
```
Please remove this change here. Currently I am checking how we will proceed with this.

schedutron commented:
@tcm-marcel I've added line numbers to the diff. And removed the stuff I've added to the gitignore file. But adding the build stuff helped in ignoring the binaries. Why shouldn't they be in .gitignore?

tcm-marcel commented:
@schedutron Can you add `.strip()` to the string before printing, to avoid double new lines? And print the file path only once per file. (I tried to add it myself, but couldn't figure out how to push to your branch. )

Regarding the ignore file: I can't see your changes anymore because the commit was removed. But even if the changes make sense, this should be a seperate PR.

tcm-marcel commented:
@schedutron Can you add `.strip()` to the string before printing, to avoid double new lines? And print the file path only once per file. (I tried to add it myself, but couldn't figure out how to push to your branch. )

Regarding the ignore file: I can't see your changes anymore because the commit was removed. But even if the changes make sense, this should be a seperate PR.

schedutron commented:
I've added the code to print file path only once. Have a look :)

tcm-marcel commented:
You added it for `def check_common_patterns(file_path)`, add this also for `def check_format(file_path)`. 
And please add `.strip()` in this line https://github.com/schedutron/peloton/blob/19652ac7bace7522f6ae86334de3f91a55d9decb/script/validators/source_validator.py#L146 to avoid the double new lines: `LOG.info("Line %d: %s" % (line_num, line[2:].strip()))`

schedutron commented:
Done.

tcm-marcel commented:
Great, thank you @schedutron!

apavlo commented:
I will merge this next when the build passes. It got caught on the LLVM download bug.
